Feature/task 1

HanaNavratilova opened the pull request.

Commit:
KA-324 add ListController

Commit:
KA-324 add list controller implementation and tests

Commit:
KA324 edit AssemblyInfo

Commit:
KA-324 add appveyor config file

Commit:
KA-324 add versioning

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/App_Start/WebApiConfig.cs:
```diff
@@ -10,7 +10,7 @@ public static void Register(HttpConfiguration config)
 
             config.Routes.MapHttpRoute(
```
You are mixing two ways how to handle versions, please look [here](https://github.com/Microsoft/aspnet-api-versioning/wiki/Versioning-via-the-URL-Path) how to handle URL path versioning. It will show you, that you do not need MapHttpRoute at all.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Controllers/ListController.cs:
```diff
@@ -0,0 +1,62 @@
+        public async Task<IHttpActionResult> PostAsync(ListItem item)
+        {
+            items.Add(item);
+            return await Task.FromResult((IHttpActionResult)Ok());
```
POST should return status code that represents [item creation](https://tools.ietf.org/html/rfc2616#section-9.5).

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Controllers/ListController.cs:
```diff
@@ -0,0 +1,62 @@
+        public async Task<IHttpActionResult> DeleteAsync(Guid id)
+        {
+            items.Remove(items.Find(item => item.Id == id));
+            return await Task.FromResult((IHttpActionResult)Ok());
```
Well it your case you should return [NoContent](https://tools.ietf.org/html/rfc2616#section-9.7).

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Controllers/ListController.cs:
```diff
@@ -0,0 +1,62 @@
+﻿using System;
```
Remove unused usings.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Models/ListItem.cs:
```diff
@@ -0,0 +1,62 @@
+﻿using System;
```
Remove unused usings.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Models/ListItem.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace MyPerfectOnboarding.Api.Models
+{
+    [DataContract]
```
You do not have to specify DataContract as you [do not specify in the json formatter to use it](https://docs.microsoft.com/en-us/aspnet/web-api/overview/formats-and-model-binding/json-and-xml-serialization). In default WebApi uses JSON.NET.

vaclavm replied:
I am still convinced that you should remove [DataContract] and [DataMember] as it should be used with different formatter. JSON.NET can work with it but in the core has its own attributes. It is never good to mix two different approaches and as long as you need to format the output to JSON you should use best practices for JSON formatter.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,108 @@
+
+            List<ListItem> items;
+
+            message.TryGetContentValue<List<ListItem>>(out items);
```
Type in TryGetContentValue is redundant, do you use it 'cause of CI?

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,108 @@
+
+            var response = await _listController.DeleteAsync(id);
+
+            var message = await response.ExecuteAsync(CancellationToken.None);
```
You use `response.ExecuteAsync(CancellationToken.None);` in all of your tests. How about writing an [extension](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods) for the controller where you would execute the specified action and returned the message? Hint, function header: ` public static async Task<HttpResponseMessage> ExecuteAction<T>(...) { ... } `

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Models/ListItem.cs:
```diff
@@ -0,0 +1,62 @@
+        private DateTime _creationTime;
+        private DateTime _lastUpdateTime;
+
+        public ListItem(Guid id, string text, bool isActive, DateTime creationTime, DateTime lastUpdateTime)
```
I will leave it be this time, but I am sure you wouldn't need the constructor in the further Tasks 😉.

vaclavm reviewed (changes requested):
Hi there, this is my first review for you (I think) 🐼.

To be honest, I don't write such a long, informational reviews as Petr does. I like personal interaction, so don't hesitate to ask anything regarding this review.

Overall your code is pretty neat! I like it and I know you struggled with it. I believe you want to upgrade your server to another level, so let's check 'theme dope comments' from me!

Commit:
KA-324 code refactoring
return correct post and delete status code
remove unused usings
modify api config: versioning, json data contract
add execute extension in tests

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/packages.config:
```diff
@@ -1,8 +1,12 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Climax.Web.Http" version="1.4.0" targetFramework="net46" />
```
Why do you have this installed?

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Models/ListItem.cs:
```diff
@@ -0,0 +1,62 @@
+namespace MyPerfectOnboarding.Api.Models
+{
+    [DataContract]
+    public class ListItem
```
Your model should be just data transfer model, you do not need constructor and any private fields. All you need is plain object with properties - after you remove constructor, even displaying fields in browser will work.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/App_Start/WebApiConfig.cs:
```diff
@@ -1,18 +1,25 @@
-                defaults: new { id = RouteParameter.Optional }
-            );
+            var json = GlobalConfiguration.Configuration.Formatters.JsonFormatter;
+            json.UseDataContractJsonSerializer = true;
```
Please, do not use data contract for JSON serialization, it is like 'drbat se levou rukou za pravym uchem'. WebAPI uses JSON.NET as default serializer, all you need to do is specify SerializerSettings.ContractResolver to camelCase because you want it to transfer your properties to camelCase (Id => id, CreationTime => creationTime), so you can connect your FE to this BE out of the box.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/App_Start/WebApiConfig.cs:
```diff
@@ -1,18 +1,25 @@
-                routeTemplate: "api/{controller}/{id}",
-                defaults: new { id = RouteParameter.Optional }
-            );
+            var json = GlobalConfiguration.Configuration.Formatters.JsonFormatter;
```
Move setting the json to new file, something like JsonFormatterConfig would be appropriate. It should work same as WebApiConfig - it will have Register method with HttpConfiguration config and will be called from same place as WebApiConfig is called.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Controllers/ListController.cs:
```diff
@@ -0,0 +1,58 @@
+    {
+        public List<ListItem> items = new List<ListItem>
+        {
+            new ListItem(Guid.NewGuid(), "aaaa", false, DateTime.MinValue, DateTime.MinValue),
```
After you remove the constructor from ListItem, you can easily change this to object initialization using { }.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Utils/Extensions.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        public static async Task<HttpResponseMessage> ExecuteAction(this IHttpActionResult response)
+        {
+            return await response.ExecuteAsync(CancellationToken.None);
```
You can take this a little bit further! How about [executing the action](https://docs.microsoft.com/en-us/dotnet/api/system.func-2?view=netframework-4.7.2) from controller here?
Therefore in the tests you would only call:
`
var response = await _listController.ExecuteAction(controller => controller.GetAsync());
`
instead of having two (almost)identical lines of code in every test:
````
var response = await _listController.GetAsync();
var message = await response.ExecuteAction();
````

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Utils/Extensions.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace MyPerfectOnboarding.Api.Tests.Utils
+{
+    public static class Extensions
```
Extensions is generic name. What is the extension used for - that should be described in the name.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,107 @@
+
+            var id = _listController.items.ElementAt(index).Id;
+
+            var response = await _listController.DeleteAsync(id);
```
For this dummy controller, it is enough to send any id in the method. Like: "aaaa" would be appropriete, the Id doesn't have to exist right now.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,107 @@
+        {
+            var index = 1;
+
+            var id = _listController.items.ElementAt(index).Id;
```
You don't have to do this, should be dummy test to correctly test the status code.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,107 @@
+
+            Assert.That(message.StatusCode, Is.EqualTo(HttpStatusCode.OK));
+
+            Assert.That(item, Is.InstanceOf<ListItem>());
```
This assert (for instance) one is unnecessary.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,107 @@
+
+            Assert.That(message.StatusCode, Is.EqualTo(HttpStatusCode.OK));
+
+            Assert.That(items, Is.InstanceOf<List<ListItem>>());
```
You don't have to test if instance is correct

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,107 @@
+        {
+            Guid id = _listController.items.First().Id;
+
+            var response = await _listController.GetAsync(id);
```
Again, it is a dummy test. Send just plain Id string like you have in the items ("aaaa").

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Controllers/ListController.cs:
```diff
@@ -0,0 +1,58 @@
+    [Route("api/v{version:apiVersion}/List")]
+    public class ListController : ApiController
+    {
+        public List<ListItem> items = new List<ListItem>
```
You really don't need this in tests, therefore you can make it private.

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Controllers/ListController.cs:
```diff
@@ -0,0 +1,58 @@
+        [HttpGet]
+        public async Task<IHttpActionResult> GetAsync()
+        {
+            return await Task.FromResult((IHttpActionResult)Ok(items));
```
Why do you cast the Ok(items) to IHttpActionResult?
You do it in every method, but it is not needed ...

vaclavm commented on MyPerfectOnboardingApplication/MyPerfectOnboarding.Api/Controllers/ListController.cs:
```diff
@@ -0,0 +1,58 @@
+        };
+
+        [HttpGet]
+        public async Task<IHttpActionResult> GetAsync()
```
The method can be made 'lamdish'.

vaclavm reviewed (changes requested):
Hi Hani, I made a deeper review now. I am sure you will be happy as it teaches you new stuff, https://some.ly/uMUpz8B/! 👼

Overall, your solution is good and works. You only mix multiple approaches together, which can be a little bit confusing. So, let's make the solution clear!

Have a good day, Vasek!

Commit:
KA-324 code refactoring
make controller more dummy
improve execute extension
add json serializer config
delete unused things in List Item class
